File: cpp/Trie.cc
#include <unordered_map>
#include <string>
#include <iostream>
using namespace std;

// Implemented using pointers similar to what's learned in CS138
// Will redo later with OOP

struct Node {
  unordered_map<char, Node*> children;
  bool eow = false;
  char val;
};

struct Trie {
  Node* root;
};

// Forward declarations
void initTrie(Trie& t);
void addWord(Trie& t, string word);
void add(Node* n, string word);
bool containsWord(Trie& t, string word);
bool contains(Node* n, string word);
int countWord(Trie& t, string word);
int countChildren(Node* n);

// Functions
void initTrie(Trie& t){
  t.root = new Node;
  t.root -> val = '\0';
}

void addWord(Trie& t, string word){
  add(t.root, word);
}

void add(Node* n, string word){
  if(word.length() == 0){
    n -> eow = true;
    return;
  } else {
    if(n -> children.find(word.at(0)) == n -> children.end()){
      n -> children[word.at(0)] = new Node;
      n -> children[word.at(0)] -> val = word.at(0);
    }
    add(n -> children[word.at(0)], word.substr(1, word.length() - 1));
  }
}

bool containsWord(Trie& t, string word){
  return contains(t.root, word);
}

bool contains(Node* n, string word){
  if(word.length() == 0){
    return true;
  } else {
    if(n -> children.find(word.at(0)) == n -> children.end()){
      return false;
    } else {
      return contains(n -> children[word.at(0)], word.substr(1, word.length() - 1));
    }
  }
}

int countWord(Trie& t, string word){
  Node* n = t.root;
  for(int i = 0; i < word.length(); i++){
    if(n -> children.find(word.at(i)) == n -> children.end()){
      return 0;
    } else {
      n = n -> children[word.at(i)];
    }
  }
  return countChildren(n);
}

int countChildren(Node* n){
  int sum = 0;
  if(n -> children.size() == 0){
    sum += 1;
  } else {
    if(n -> eow){
      sum += 1;
    }
    for(auto it = n -> children.begin(); it != n -> children.end(); ++it){
      sum += countChildren(it -> second);
    }
  }
  return sum;
}

int main(){
  Trie t;
  initTrie(t);
  addWord(t, "cat");
  addWord(t, "card");
  addWord(t, "car");
  addWord(t, "dog");
  cout << countWord(t, "ca") << endl;
}

File: python/Trie.py
class Node:
    def __init__(self, val):
        self.children = {}
        self.eow = False
        self.val = val


class Trie:
    def __init__(self):
        self.root = Node("")

    def addWord(self, word):
        self._add(word, self.root)

    def _add(self, word, node):
        if len(word) == 0:
            node.eow = True
            return
        else:
            if node.children.get(word[0]) is None:
                node.children[word[0]] = Node(word[0])
            self._add(word[1:], node.children[word[0]])

    def contains(self, word):
        return self._contains(word, self.root)

    def _contains(self, word, node):
        if len(word) == 0:
            return True
        else:
            if node.children.get(word[0]) is None:
                return False
            else:
                return self._contains(word[1:], node.children[word[0]])

    def countWords(self, word):
        if not self.contains(word):
            return 0
        else:
            node = self.root
            for c in word:
                node = node.children[c]
            return self._countChildren(node)

    def _countChildren(self, node):
        sum = 0

        if len(node.children.keys()) == 0:
            sum += 1
        elif len(node.children.keys()) != 0:
            if node.eow:
                sum += 1
            for n in node.children:
                sum += self._countChildren(node.children[n])

        return sum